File: a_backend/src/services/geminiService.js
const { GoogleGenerativeAI } = require('@google/generative-ai');
const { formatSize } = require('../utils/helpers');

const genAI = new GoogleGenerativeAI(process.env.GEMINI_API_KEY);

async function searchFiles(query, filesWithContent) {
  if (filesWithContent.length === 0) return [];

  const model = genAI.getGenerativeModel({ model: 'gemini-1.5-flash' });

  const fileContext = filesWithContent
    .map((file) => {
      let block = `=== ARQUIVO: ${file.name} | Tipo: ${file.ext || 'desconhecido'} | Tamanho: ${formatSize(file.size)} ===\n`;
      block += file.content
        ? file.content
        : '[Arquivo binário — conteúdo não extraível, avalie pelo nome/extensão]';
      return block;
    })
    .join('\n\n---\n\n');

  const prompt = `Você é um assistente de busca de arquivos altamente preciso. Analise os arquivos listados abaixo e identifique TODOS que têm alguma relação com a descrição do usuário. Seja inclusivo: na dúvida, inclua o arquivo.

DESCRIÇÃO DO USUÁRIO: "${query}"

ARQUIVOS DISPONÍVEIS:
${fileContext}

REGRAS:
- Analise tanto o nome quanto o conteúdo de cada arquivo.
- Inclua arquivos com relação direta E indireta com a descrição.
- Escreva os motivos em português, de forma breve e clara.
- Responda SOMENTE com JSON válido — sem markdown, sem explicações fora do JSON.

FORMATO DE RESPOSTA:
{"relevant_files":[{"name":"nome_exato_do_arquivo.ext","reason":"Motivo breve em português"}]}

Se nenhum arquivo for relevante: {"relevant_files":[]}`;

  const result = await model.generateContent(prompt);
  const raw = result.response.text().trim();

  // Remove blocos de código markdown se presentes
  let jsonStr = raw.replace(/```json\s*/g, '').replace(/```\s*/g, '').trim();

  // Extrai o objeto JSON da resposta
  const match = jsonStr.match(/\{[\s\S]*\}/);
  if (match) jsonStr = match[0];

  try {
    const parsed = JSON.parse(jsonStr);
    return Array.isArray(parsed.relevant_files) ? parsed.relevant_files : [];
  } catch (err) {
    console.error('Falha ao parsear resposta do Gemini:', raw);
    return [];
  }
}

module.exports = { searchFiles };

File: a_backend/src/services/gemini_service.py
import os
import re
import json
import asyncio
from typing import Any

from google import genai

from src.utils.helpers import format_size
from src.services.vector_service import search as vector_search

_client = genai.Client(api_key=os.getenv("GEMINI_API_KEY"))


async def search_files(
    query: str,
    all_files: list[dict[str, Any]],
    user_id: int,
) -> list[dict[str, str]]:
    """
    Busca arquivos relevantes para `query` usando RAG:
      1. ChromaDB recupera os chunks mais similares semanticamente
      2. Gemini avalia os candidatos e retorna os relevantes com motivo
    """
    if not all_files:
        return []

    # ── Etapa 1: RAG — chunks relevantes ────────────────────────────────────
    chunks = await vector_search(query, user_id, n_results=12)

    # Monta conjunto de arquivos candidatos (únicos) com conteúdo relevante
    seen: set[str] = set()
    candidates: list[dict] = []
    for chunk in chunks:
        key = f"{chunk['folder']}|{chunk['filename']}"
        if key not in seen:
            seen.add(key)
            candidates.append(chunk)

    # Complementa com arquivos binários/não-indexados (avaliados pelo nome)
    indexed_names = {c["filename"] for c in candidates}
    for f in all_files:
        if f["name"] not in indexed_names:
            candidates.append({
                "filename": f["name"],
                "folder":   f.get("folder") or "",
                "content":  None,
                "score":    0,
            })

    if not candidates:
        return []

    # ── Etapa 2: Gemini avalia os candidatos ─────────────────────────────────
    context_parts = []
    for c in candidates:
        loc = f" (em: {c['folder']}/)" if c["folder"] else ""
        header = f"=== {c['filename']}{loc} ==="
        body   = c["content"] or "[sem conteúdo textual — avalie pelo nome/extensão]"
        context_parts.append(f"{header}\n{body}")
    file_context = "\n\n---\n\n".join(context_parts)

    prompt = f"""Analise os arquivos abaixo e identifique os relacionados com a descrição.

DESCRIÇÃO: "{query}"

ARQUIVOS CANDIDATOS:
{file_context}

RETORNE somente JSON válido:
{{"relevant_files":[{{"name":"nome.ext","reason":"motivo em português"}}]}}

Se nenhum for relevante: {{"relevant_files":[]}}"""

    response = await asyncio.to_thread(
        _client.models.generate_content,
        model="gemini-2.5-flash",
        contents=prompt,
    )
    raw = response.text.strip()
    json_str = re.sub(r"```[a-z]*\s*", "", raw).replace("```", "").strip()
    match = re.search(r"\{[\s\S]*\}", json_str)
    if match:
        try:
            return json.loads(match.group(0)).get("relevant_files", [])
        except json.JSONDecodeError:
            pass
    return []
